### PR TITLE
Fleshes out more of paying vote rewards for Alpenglow

### PR DESCRIPTION
**Pull request by akhi3030**
#### Problem

Previously, we were just calculating the vote rewards but not actually paying them.  This PR fleshes out more of paying vote rewards.  More specifically, we pay the vote rewards into the vote accounts and also update account state to aid delinquency monitoring.

#### Summary of Changes

- Fleshes out the data we need to store in the off curve account.  In particular, we need to store info for 2 epochs and additional information on top of validator rewards.
- We pay the vote rewards in the vote accounts and additionally update the delinquency list based on which validators signed the final cert.
- Increases the test coverage

### CODE REVIEW

File: runtime/src/block_component_processor/vote_reward.rs [hunk 1]
```diff
@@ -1,189 +1,408 @@
...
+        latency: 0,
+    }]);
+}
+
+#[derive(Debug, Error)]
+enum ConvertError {
+    #[error("failed to find node_vote_accounts for {node_pubkey}")]
+    NodeVoteAccountsNotFound { node_pubkey: Pubkey },
+    #[error("node pubkey should map to exactly {expected} vote account but maps to {got} accounts")]
+    VoteAccountsLenMismatch { expected: usize, got: usize },
```
Comment (AshwinSekar) [inline runtime/src/block_component_processor/vote_reward.rs]: I don't  think we need `expected` here, it should always be 1.

Also would be good to have a test where we have multiple vote accounts for the same node pubkey and we show that rewards are dropped.

**Comment by akhi3030** [comment]
@AshwinSekar : as discussed offline, I uploaded my debugging code in https://github.com/anza-xyz/agave/pull/11434/commits/bf4c563b92636e879716fb42d4d4aee4f4f49a85.

It adds a new test: `leader_with_multiple_vote_accounts_not_paid`.  To run, you need to do: `RUST_BACKTRACE=1 cargo nextest run --no-capture -- leader_with_multiple`

I get the following output:

```
handling vote account pubkey=9XGftH26VSZ4mekqNCU3UUXkj5ewkyQtg8FGWg6GCWYd
process_new_epoch: vote_accounts = [9XGftH26VSZ4mekqNCU3UUXkj5ewkyQtg8FGWg6GCWYd]
handling vote account pubkey=9XGftH26VSZ4mekqNCU3UUXkj5ewkyQtg8FGWg6GCWYd
calc: reward_slot_accounts = [4KXwfnqLwjTa9H72tnxjcYxKCeb99N1QMKkfpYfjqkq9, 9GkAMbLdxZNgfvUtQhew8wR4bUDj2Qdf7Pf6dFqxa66T, 8bQ5iu5HRjU4oS3afU8woABmUWsDS3Bgfc9yNdJ1DKA9, AtNeHpDQpzoNoR9toNMLnwHSEKGANSqdgaeriRykxcdC]
add=9XGftH26VSZ4mekqNCU3UUXkj5ewkyQtg8FGWg6GCWYd epoch_credits=[]
```

In the test, I use `new_vote_accounts(1, 4, true);` so I should get 4 vote accounts.  And I also created one more when I created the genesis config.  Hence, there should be a total 5 vote accounts.  However, the print in `calculate_and_pay_voting_reward_and_update_vote_state` shows just 4.  This kinda makes sense as the first account (created by the genesis config) is in `Bank::stakes_cache` and the other 4 are in `VersionedEpochStakes`.

The final print shows that we iterated over just a single vote account which was the one stored in `Bank::stakes_cache`.

Also note that in the test I create the vote accounts on `prev_bank` and then I force a epoch boundary crossing when I create `bank`.  So `process_new_epoch` should be getting called.  Looking at the code of `process_new_epoch`, it seems like it is just looking at the parent's `stakes_cache` to populate the child's `stakes_cache`.

File: runtime/src/block_component_processor/vote_reward.rs [hunk 2]
```diff
@@ -1,189 +1,417 @@
...
+            *reward_slot_validator_stake,
         );
-        // As per the Alpenglow SIMD, the rewards are split equally between the validators and the leader.
-        let validator_reward_lamports = reward_lamports / 2;
-        let leader_reward_lamports = reward_lamports - validator_reward_lamports;
-        total_leader_reward_lamports =
-            total_leader_reward_lamports.saturating_add(leader_reward_lamports);
+        total_leader_reward = total_leader_reward.saturating_add(add_leader_reward);
+
+        // assuming that the size of the this vec is small, the O(N) lookup here should not be expensive.
```
Comment (akhi3030) [inline runtime/src/block_component_processor/vote_reward.rs]: @AshwinSekar: do you think we need to worry about the O(N) lookup here?  A malicious actor could setup accounts to abuse this but is that a realistic attack?  I can always convert this Vec to a HashSet just wondering if that is actually useful in practice.

File: runtime/src/block_component_processor/vote_reward.rs (hunk 1 above)
Comment (akhi3030) [reply runtime/src/block_component_processor/vote_reward.rs]: Both points should now be addressed.

**Comment by akhi3030** [comment]
In developing a test for when a leader has multiple vote accounts, I found an issue in the code where we were still paying one of the leader's account even when it had multiple vote accounts.  I had to change the code around a bit to handle that.

File: runtime/src/block_component_processor/vote_reward/epoch_inflation_account_state.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    pub(crate) fn new_from_bank(bank: &Bank) -> Option<Self> {
+        bank.get_account(&VOTE_REWARD_ACCOUNT_ADDR)
+            .and_then(|acct| wincode::deserialize(acct.data()).ok())
+    }
+
+    /// Serializes and updates [`Self`] into the accounts in the [`Bank`].
+    pub(super) fn set_state(&self, bank: &Bank) {
+        let account_size = wincode::serialized_size(&self).unwrap();
+        let lamports = bank.get_minimum_balance_for_rent_exemption(account_size as usize);
+        let account = AccountSharedData::new_data(lamports, &self, &system_program::ID).unwrap();
```
Comment (AshwinSekar) [inline runtime/src/block_component_processor/vote_reward/epoch_inflation_account_state.rs]: we can force wincode using this pattern
https://github.com/anza-xyz/agave/blob/b55d8fd61ad15727e0d574470e09e6767eb01a6e/runtime/src/bank.rs#L3054-L3060

File: runtime/src/block_component_processor/vote_reward.rs [hunk 4]
```diff
@@ -1,266 +1,709 @@
...
+    // As per the Alpenglow SIMD, the rewards are split equally between the validators and the leader.
+    let validator_reward_lamports = reward_lamports / 2;
+    let leader_reward_lamports = reward_lamports - validator_reward_lamports;
+    (validator_reward_lamports, leader_reward_lamports)
+}
+
+/// Deserializes `VoteState` from `account`; pays `reward` in `current_epoch` to the `epoch_credits` field;
+/// and updates the `votes` and `root_slot` fields in the vote state deserialized from the `account`.
+///
+/// TODO: this is using VoteStateV4 explicitly.  When we upstream, we will use VoteStateHandle API.
```
Comment (AshwinSekar) [inline runtime/src/block_component_processor/vote_reward.rs:210]: do we want to use VoteStateHandle now?

File: runtime/src/block_component_processor.rs
Comment (AshwinSekar) [inline runtime/src/block_component_processor.rs]: do we still need this fn now that we have `update_bank_with_footer_fields`?

File: runtime/src/block_component_processor/vote_reward/epoch_inflation_account_state.rs (hunk 3 above)
Comment (akhi3030) [reply runtime/src/block_component_processor/vote_reward/epoch_inflation_account_state.rs]: nice!  Done

File: runtime/src/block_component_processor.rs
Comment (akhi3030) [reply runtime/src/block_component_processor.rs]: good catch, this is dead code now.

File: runtime/src/block_component_processor/vote_reward.rs (hunk 4 above)
Comment (akhi3030) [reply runtime/src/block_component_processor/vote_reward.rs:210]: I am hoping to land this and work on a separate PR to migrate to `VoteStateHandle`.  Created https://github.com/anza-xyz/agave/issues/11585 for it as well.

**Review by AshwinSekar — APPROVED** [review]
